File: leet/1345_jump_game_4/1345_jump_game_4.py
# LeetCode 1345. Jump Game IV
# https://leetcode.com/problems/jump-game-iv/
# O(n) time O(n) space
class Solution:
    def minJumps(self, arr: List[int]) -> int:
        edges = defaultdict(list)
        for i, n in enumerate(arr):
            edges[n].append(i)
        steps, queue, visited = 0, deque(), set()
        queue.append(0)
        while len(queue):
            for _ in range(len(queue)):
                i = queue.popleft()
                if i == len(arr) - 1:
                    return steps
                visited.add(i)
                if i > 0 and i - 1 not in visited:
                    queue.append(i - 1)
                if i < len(arr) - 1 and i + 1 not in visited:
                    queue.append(i + 1)
                if arr[i] in edges:
                    for j in edges[arr[i]]:
                        if j not in visited:
                            queue.append(j)
                    del edges[arr[i]]
            steps += 1
        return steps

File: leet/1345_jump_game_4/1345_jump_game_4.js
// LeetCode 1345. Jump Game IV
// https://leetcode.com/problems/jump-game-iv/
// O(n) time-and-space
var minJumps = function(arr) {
    const valToIndices = {}
    for (let i = 0; i < arr.length; i++) {
        if (!valToIndices[arr[i]])
            valToIndices[arr[i]] = [];
        valToIndices[arr[i]].push(i);
    }
    const visited = Array(arr.length).fill(false), queue = [0];
    visited[0] = true;
    let steps = 0, x = 0;
    while (queue.length) {
        for (let sz = queue.length - x; sz > 0; sz--) {
            i = queue[x++];
            if (i == arr.length - 1)
                return steps;
            visited[i] = true;
            for (j of valToIndices[arr[i]]) {
                if (!visited[j])
                    queue.push(j)
            }
            valToIndices[arr[i]] = [];
            if (i + 1 < arr.length && !visited[i + 1])
                queue.push(i + 1);
            if (i - 1 >= 0 && !visited[i - 1])
                queue.push(i - 1);
        }
        steps++;
    }
    return steps;

};
